sdk/identity/azure-identity/azure/identity/_internal/shared_token_cache.py

```diff
@@ -0,0 +1,185 @@
+SharedTokenCacheCredential, login through developer tooling supporting Azure single sign on.
+Discovered accounts: {}"""
+
+NO_TOKEN = """Token acquisition failed for user '{}'. To fix, re-authenticate
+through developer tooling supporting Azure single sign on"""
+
+_PUBLIC_CLOUD_ALIASES = {"login.windows.net", "login.microsoft.com", "sts.windows.net"}
+
+
+def _account_to_string(account):
+    username = account.get("username")
+    home_account_id = account.get("home_account_id", "").split(".")
+    tenant_id = home_account_id[-1] if len(home_account_id) == 2 else ""
+    return "(username: {}, tenant: {})".format(username, tenant_id)
+
+
+def _filtered_accounts(accounts, username=None, tenant_id=None):
+    """yield accounts matching username and/or tenant_id"""
+
+    filtered_accounts = []
+    for account in accounts:
+        if username and account.get("username") != username:
+            continue
+        if tenant_id:
+            try:
+                _, tenant = account["home_account_id"].split(".")
+                if tenant_id != tenant:
+                    continue
+            except Exception:  # pylint:disable=broad-except
+                continue
+        filtered_accounts.append(account)
+    return filtered_accounts
+
+
+class SharedTokenCacheBase(ABC):
+    def __init__(self, username=None, **kwargs):  # pylint:disable=unused-argument
+        # type: (Optional[str], **Any) -> None
+
+        self._authority = kwargs.pop("authority", KnownAuthorities.AZURE_PUBLIC_CLOUD)
+        self._username = username
+        self._tenant_id = kwargs.pop("tenant_id", None)
+
+        cache = kwargs.pop("_cache", None)  # for ease of testing
+
+        if not cache and sys.platform.startswith("win") and "LOCALAPPDATA" in os.environ:
+            from msal_extensions.token_cache import WindowsTokenCache
+
+            cache = WindowsTokenCache(cache_location=os.environ["LOCALAPPDATA"] + "/.IdentityService/msal.cache")
+
+            # prevent writing to the shared cache
+            # TODO: seperating deserializing access tokens from caching them would make this cleaner
+            cache.add = lambda *_, **__: None
+
+        if cache:
+            self._cache = cache
+            self._client = self._get_auth_client(cache=cache, **kwargs)  # type: Optional[AadClientBase]
+        else:
+            self._client = None
+
+    @abc.abstractmethod
+    def _get_auth_client(self, **kwargs):
+        # type: (**Any) -> AadClientBase
+        pass
+
+    def _get_cache_items_for_authority(self, credential_type):
+        # type: (TokenCache.CredentialType) -> List[CacheItem]
+        """yield cache items matching this credential's authority or one of its aliases"""
+
+        items = []
+        for item in self._cache.find(credential_type):
+            environment = item.get("environment")
+            if environment == self._authority or (
```

Comment:
Do other clouds have authority aliases? Or is this just a public cloud phenomenon?

Reply:
China and USG have [one alias each](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/master/src/client/Microsoft.Identity.Client/Instance/Discovery/KnownMetadataProvider.cs#L38). I was going to leave that aside, hoping that before it became a problem I'd be using MSAL to solve it, but once you pointed it out I felt nervous enough to support them now.